client/core/core.go

```diff
@@ -1999,34 +2033,105 @@ func (c *Core) Register(form *RegisterForm) (*RegisterResult, error) {
 		return nil, newError(assetSupportErr, "dex server does not support %s asset", regFeeAssetSymbol)
 	}
 
-	privKey, err := dc.acct.setupEncryption(crypter)
-	if err != nil {
-		return nil, codedError(acctKeyErr, err)
+	creds := c.creds()
+
+	// Depending on the age of the server, we'll either generate a hierarchical
+	// or a legacy key pair.
+	var acctPubB, encKey, encKeyLegacy []byte
+	if dc.acct.dexPubKey != nil {
+		encKey, acctPubB, err = dc.acct.setupCryptoV2(creds, crypter)
+		if err != nil {
+			return nil, newError(acctKeyErr, "setupCryptoV2 error: %v", err)
+		}
+
+	} else {
+		// This is an old-style DEX that doesn't provide their pubkey in the
+		// config response. Gotta use a legacy-style key.
+		encKeyLegacy, acctPubB, err = dc.acct.setupCryptoLegacy(crypter)
+		if err != nil {
+			return nil, codedError(acctKeyErr, err)
+		}
 	}
 
 	// Prepare and sign the registration payload.
-	// The account ID is generated from the public key.
-	pkBytes := privKey.PubKey().SerializeCompressed()
+	storeAccount := func() error {
+		return c.db.CreateAccount(&db.AccountInfo{
+			Host:         dc.acct.host,
+			Cert:         dc.acct.cert,
+			LegacyEncKey: encKeyLegacy,
+			EncKeyV2:     encKey,
+			DEXPubKey:    dc.acct.dexPubKey,
+			FeeCoin:      dc.acct.feeCoin,
+		})
+	}
+
 	dexReg := &msgjson.Register{
-		PubKey: pkBytes,
+		PubKey: acctPubB,
 		Time:   encode.UnixMilliU(time.Now()),
 	}
 	regRes := new(msgjson.RegisterResult)
 	err = dc.signAndRequest(dexReg, msgjson.RegisterRoute, regRes, DefaultResponseTimeout)
 	if err != nil {
-		return nil, codedError(registerErr, err)
+		var msgErr *msgjson.Error
+		if !errors.As(err, &msgErr) || msgErr.Code != msgjson.AccountExistsError {
+			return nil, codedError(registerErr, err)
+		}
+
+		// It's an AccountExistsError. This is now account recovery, which is
+		// great news since we don't have to pay the fee.
+		// DRAFT TODO: How to handle suspended accounts?
+
```

Comment:
Just want to draw attention to this remaining issue. What do we do if the recovered key is associated with a suspended account?

1. Can we allow repaying a fee on the same account ID to get unsuspended?
2. Should we just generate a random key and treat it as an imported account, requiring separate backup?

Reply:
I don't think one is good, because what if the server decides to forgive us? Two is a shame when you already got this great deterministic scheme set up.

Can add one more `kid` in `setupCryptoV2` and if told here we're suspended, increment that index. Increment until we find an account ID that is not suspended/banned.